File: packages/ts-sdk-fhir/src/builders/encounter.ts
/**
 * `Encounter` builder.
 *
 * Used by clinical modules (OPD, IPD, ER) to serialise visit/admission rows
 * into FHIR `Encounter` resources, attached to `*.finalized` event payloads
 * per ADR-0023.
 *
 * @see docs/architecture/adr/0023-distributed-fhir-assembly.md
 * @see https://hl7.org/fhir/R4/encounter.html
 */

import type {
  Encounter,
  FhirCodeableConcept,
  FhirCoding,
  FhirIdentifier,
  FhirMeta,
  FhirPeriod,
  FhirReference,
} from "../types/index.js";

// eslint-disable-next-line sonarjs/no-clear-text-protocols -- canonical FHIR system URI (identifier, not a network call)
const V3_ACT_ENCOUNTER = "http://terminology.hl7.org/CodeSystem/v3-ActCode";

export interface BuildEncounterInput {
  identifier?: FhirIdentifier[];
  status: Encounter["status"];
  /** ActEncounterCode (`AMB`, `IMP`, `EMER`, …). `system` should be v3-ActCode when set. */
  class: FhirCoding;
  type?: FhirCodeableConcept[];
  subject: FhirReference;
  period?: FhirPeriod;
  /** Optional `meta.profile` URLs (NRCeS or core StructureDefinitions). */
  meta?: FhirMeta;
}

/**
 * Build an `Encounter` resource.
 *
 * When `class.system` is set, it must be the v3 ActEncounterCode code system.
 *
 * @see docs/architecture/adr/0023-distributed-fhir-assembly.md
 */
export function buildEncounter(input: BuildEncounterInput): Encounter {
  if (input.class.system && input.class.system !== V3_ACT_ENCOUNTER) {
    throw new TypeError(
      `Encounter.class.system must be "${V3_ACT_ENCOUNTER}" when provided (got ${input.class.system})`,
    );
  }
  return {
    resourceType: "Encounter",
    ...(input.identifier ? { identifier: input.identifier } : {}),
    ...(input.meta ? { meta: input.meta } : {}),
    status: input.status,
    class: input.class,
    ...(input.type ? { type: input.type } : {}),
    subject: input.subject,
    ...(input.period ? { period: input.period } : {}),
  };
}

File: packages/py-sdk-fhir/src/hims_sdk_fhir/builders/encounter.py
"""``build_encounter`` — Layer-1 Encounter resource builder.

Emits the literal ``"class"`` JSON key (the TypedDict field is ``class_``).

@see docs/architecture/adr/0023-distributed-fhir-assembly.md
@see https://hl7.org/fhir/R4/encounter.html
"""

from __future__ import annotations

from ..inputs import EncounterInput
from ..lib import compact
from ..profile_registry import resource_profile
from ..types import Encounter, FhirReference

_V3_ACT_CODE = "http://terminology.hl7.org/CodeSystem/v3-ActCode"

# ActEncounterCode display lookup (extend as new class codes are used).
_CLASS_DISPLAY: dict[str, str] = {
    "AMB": "ambulatory",
    "IMP": "inpatient encounter",
    "EMER": "emergency",
    "HH": "home health",
    "VR": "virtual",
}


def build_encounter(
    inp: EncounterInput,
    *,
    resource_id: str,
    subject: FhirReference,
    now: str,
) -> Encounter:
    """Build an ``Encounter`` resource.

    Identifier value falls back to ``resource_id`` when no ``visit_number``;
    ``class`` is the v3 ActCode coding for ``class_code``; ``period.start``
    defaults to ``now``.
    """
    encounter: Encounter = {
        "resourceType": "Encounter",
        "id": resource_id,
        "meta": {"profile": [resource_profile("Encounter")]},
        "identifier": [{"system": "https://ndhm.in", "value": inp.visit_number or resource_id}],
        "status": inp.status,
        # Literal "class" key — see types.Encounter.class_ note.
        "class": {
            "system": _V3_ACT_CODE,
            "code": inp.class_code,
            "display": _CLASS_DISPLAY.get(inp.class_code, inp.class_code),
        },
        "subject": subject,
        "period": {"start": inp.start or now},
    }
    return compact(encounter)
